File: code/168-Excel Sheet Column Title/168-Excel Sheet Column Title.py
# -*- coding: utf-8 -*
"""
LeetCode 168 - Excel Sheet Column Title

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
"""
r"""
@author: Kai Chen
@file: 168-Excel Sheet Column Title.py
@time: 2020/7/27 10:07
@desc:
Given a positive integer, return its corresponding column title as appear in an Excel sheet.

For example:

    1 -> A
    2 -> B
    3 -> C
    ...
    26 -> Z
    27 -> AA
    28 -> AB
    ...
Example 1:

Input: 1
Output: "A"
Example 2:

Input: 28
Output: "AB"
Example 3:

Input: 701
Output: "ZY"
"""


class Solution(object):
    def convertToTitle(self, n):
        """
        :type n: int
        :rtype: str
        """
        Alphabet = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                    'U', 'V', 'W', 'X', 'Y', 'Z']
        res = []
        while n > 0:
            n -= 1
            c = n % 26
            res.insert(0, Alphabet[c])
            n = n // 26
        return "".join(res)


if __name__ == '__main__':
    s = Solution()
    print(s.convertToTitle(28))

File: code/168-Excel Sheet Column Title/168-Excel Sheet Column Title.cpp
/*
LeetCode 168 - Excel Sheet Column Title

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
*/
/*
Given an integer columnNumber, return its corresponding column title as it appears in an Excel sheet.

For example:

A -> 1
B -> 2
C -> 3
...
Z -> 26
AA -> 27
AB -> 28
...


Example 1:

Input: columnNumber = 1
Output: "A"
Example 2:

Input: columnNumber = 28
Output: "AB"
Example 3:

Input: columnNumber = 701
Output: "ZY"
Example 4:

Input: columnNumber = 2147483647
Output: "FXSHRXW"


Constraints:

1 <= columnNumber <= 231 - 1
*/
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string convertToTitle(int columnNumber) {
        string res = "";
        while (columnNumber) {
            columnNumber --;
            res += 'A' + columnNumber % 26;
            columnNumber /= 26;
        }
        reverse(res.begin(), res.end());
        return res;
    }
};
